### PR TITLE
Upload image to IPFS, pin it and construct a metadata objects and then upload and pin this JSON object for final metadata IPFS url

### PR DESCRIPTION
**Pull request by Yuripetusko**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -265,3 +265,9 @@ When running a local chain, you can run `yarn seed` to populate the chain with p
 You can see how the seeders are written in `test/seed/default`. `yarn seed` will by default execute all the seeds in the `default` folder. If you want to execute only your own seeders, put them into a subfolder inside `test/seed` and provide the folder name: `yarn seed myfolder`.
 
 Check that all edge cases are covered by running [Consolidate](#consolidate).
+
+## Generate Metadata
+This scripts generates an array of objects with metadata IPFS urls ready to be added to NFTs.
+First create a seed json file with an array of metadata fields and file path (see `metadata-seed.example.json` for example). This script will first upload image to IPFS and pin it using pinata and then upload metadata JSON object to IPFS and pin it, and return array of IPFS urls with metadata JSON ready to be added to NFTs
+
+`PINATA_KEY=XXX PINATA_SECRET=XXX yarn cli:metadata --input=metadata-seed.example.json --output=metadata-seed-outpute.json`
```
Comment (Swader) [inline README.md:273]:
```suggestion
## Generate Metadata

This script generates an array of objects with metadata IPFS URIs ready to be added to NFTs.

First, create a seed JSON file with an array of metadata fields and file path (see `metadata-seed.example.json` for example) for each image. This script will first upload the image to IPFS and pin it using [Pinata](https://pinata.cloud) and then upload the metadata JSON object to IPFS and pin it, returning an array of IPFS urls ready to be added to NFTs and/or collections.


    PINATA_KEY=XXX PINATA_SECRET=XXX yarn cli:metadata --input=metadata-seed.example.json --output=metadata-seed-output.json


> Note that it is recommended to pin the resulting hashes into multiple additional pinning services or (better) your own IPFS node to increase dissemination of the content.
```

File: cli/metadata.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  );
+  const metadataInput = JSON.parse(inputFile.toString());
+
+  if (!metadataInput?.metadata || !Array.isArray(metadataInput?.metadata)) {
+    throw new Error("Metadata json file missing 'metadata' array");
+  }
+
+  const promises = (metadataInput.metadata as MetadataSeedItem[]).map(
+    async (metadata, index) => {
+      if (!metadata.imagePath) {
```
Comment (Swader) [inline cli/metadata.ts:31]: Also check if image is readable before attempting upload or doing anything else. In fact, it might be good to go through all the imagePath entries in the JSON first and cancel whole process if even one of them is missing / unreadable, so warn the user who's about to upload a ton of meta.

File: src/tools/metadata-to-ipfs.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import fs from "fs";
+// @ts-ignore
```
Comment (Swader) [inline src/tools/metadata-to-ipfs.ts:2]: Why?

File: src/tools/metadata-to-ipfs.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import pinataSDK from "@pinata/sdk";
+import { NFTMetadata } from "../rmrk1.0.0/classes/nft";
+
+const pinata = pinataSDK(process.env.PINATA_KEY, process.env.PINATA_SECRET);
+
+const defaultOptions = {
+  pinataOptions: {
+    cidVersion: 1,
```
Comment (Swader) [inline src/tools/metadata-to-ipfs.ts:10]: Is this needed of can we let Pinata decide by default and pick whatever is current?

File: src/tools/metadata-to-ipfs.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  },
+};
+
+export const pinToIpfs = async (filePath: string, name?: string) => {
+  const options = { ...defaultOptions, pinataMetadata: { name } };
+  try {
+    const readableStreamForFile = fs.createReadStream(filePath);
```
Comment (Swader) [inline src/tools/metadata-to-ipfs.ts:17]: So this will crash if image is not readable or present, as stated above. Check for it please.

**Review by Swader — CHANGES_REQUESTED** [review]

File: src/tools/metadata-to-ipfs.ts (hunk 3 above)
Comment (Yuripetusko) [reply src/tools/metadata-to-ipfs.ts:2]: Yeah unfortunalty pinata-sdk doesn't have exported types not in this package not in any community packages :(

**Comment by Swader** [comment]
Making updates to seeder on this branch to keep in sync with the metadata factory. Shouldn't conflict.

File: src/tools/metadata-to-ipfs.ts (hunk 4 above)
Comment (Yuripetusko) [reply src/tools/metadata-to-ipfs.ts:10]: I read here that cid version 1 is new standard and should be preferred now so I decided to stick with it:

https://docs.ipfs.io/concepts/content-addressing/

File: cli/metadata.ts (hunk 2 above)
Comment (Yuripetusko) [reply cli/metadata.ts]: Added validation

File: src/tools/metadata-to-ipfs.ts (hunk 5 above)
Comment (Yuripetusko) [reply src/tools/metadata-to-ipfs.ts:17]: Added validation

**Comment by Swader** [comment]
Something fails with the image check

image

image

image
